lib/openhab/dsl/items/group_item.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+require 'forwardable'
+require 'java'
+require 'openhab/dsl/items/item_command'
+require 'openhab/dsl/items/item_delegate'
+require 'openhab/core/entity_lookup'
+
+module OpenHAB
+  module DSL
+    module Items
+      #
+      # Delegator to OpenHAB Group Item
+      #
+      class GroupItem
+        extend OpenHAB::DSL::Items::ItemCommand
+        extend OpenHAB::DSL::Items::ItemDelegate
+        include Enumerable
+
+        def_item_delegator :@group_item
+
+        #
+        # Indicator struct interpreted by rules to trigger based on items contained in a group
+        #
+        GroupItems = Struct.new(:group, keyword_init: true)
+
+        #
+        # Create a new GroupItem
+        #
+        # @param [Java::Org::openhab::core::items::GroupItem] group_item OpenHAB GroupItem to delegate to
+        #
+        def initialize(group_item)
+          @group_item = group_item
+
+          item_missing_delegate { @group_item }
+          add_state_methods
+          add_command_methods
+        end
+
+        #
+        # Iterates through the direct members of the Group
+        #
+        def each(&block)
+          OpenHAB::Core::EntityLookup.decorate_items(@group_item.members.to_a).each(&block)
+        end
+
+        #
+        # Get all members of the group recursively
+        #
+        # @return [Array] An Array containing all descendants of the Group
+        def all_members
+          OpenHAB::Core::EntityLookup.decorate_items(@group_item.all_members.to_a)
+        end
+
+        def items
+          GroupItems.new(group: self)
```

Comment:
While we're at it, should we change this to `members`, and alias `items` to it, for backwards compatibility?

`members` should also be an array of direct descendants, in contrast to all_members.

It would read better in the rule as `changed Switches.members` rather than Switches.items.

Reply:
The problem is that `members` is used by the each-method to get the group members. The items is (i believe?) only ment to be used in triggers to differentiate between member of-triggers and item triggers when passing a group item.

Reply:
Is `members` already delegated to group_item? In any case, can we somehow also use that for member-of triggers thus making it dual purpose?

So that `changed Switches` is for a trigger on the GroupItem itself, and `changed Switches.members` is for a trigger on the members of the group (ie member-of trigger)

Reply:
PS I should be commenting on the `def items` line instead.

Reply:
If `members` is used in the trigger I think it would expand to a list of items, which would create a separate trigger for all the items. But this wouldn't include any new items added to the group later. So we need some way to differentiate between:

1. Changes/updates to the actual group item's state (itemState*Trigger with the group as item)
2. Changes to one of the group's members (groupState*Trigger)
3. Expanding to a list of items for which to create individual triggers (multiple itemState*Trigger)

Not sure if 3 is needed, but that is the current consequence of using `members` in the trigger.

Reply:
What @pacive outlines is the core problem..  Differentiating between when it should be a group trigger or an item trigger.  I considered a few possibilities and all of them have drawbacks.
One option is to have member_updated, member_changed, as well as updated and changed, instead but I wasn't particularly fond of that and it gets messy with received_command too.

another options is
changed Group, :members

Reply:
https://github.com/pacive/openhab-jruby/pull/1

Reply:
Created a trial with using `changed Group, :members` (only for the changed trigger so far). Let me know what you think.

Reply:
I think the syntax `changed Group.members` looks better / less confusing than `changed Group, :members`.

Consider:
```
changed Group, :members, Item1, Item2
vs
changed Group.members, Item1, Item2
```
The :members syntax looks a bit odd because it's not a part of Group in the definition.

Reply:
I can agree with that, but can't see how we can work around the issue of using this both for signaling a group trigger and using it for getting an Array of the members. (Without resorting to the same technique of using a Struct or mediating class as the previous implementation)

Reply:
I've just added MembersOfGroup.group to refer back to the group_item

Reply:
@pacive Thanks for exploring this further. I see the problem with Group.members, and not sure how to get around the flatten issue.

How about `changed Item2, Item3, member_of: Group, to: XX`
`changed Item2, Item3, member_of: [Group1, Group2]`

It's not ideal either, but an alternative to `Group, :members` syntax to consider.

Reply:
Or we can stick to Group.items like before - less breaking change

Reply:
That would work, but could only be used once. `changed Item1, member_of: Group1, member_of: Group2` wouldn't work. Another approach would be to make it a method: `changed Item1, member_of(Group1), member_of(Group2)`, but it doesn't read as nice.

Reply:
> That would work, but could only be used once. `changed Item1, member_of: Group1, member_of: Group2` wouldn't work. Another approach would be to make it a method: `changed Item1, member_of(Group1), member_of(Group2)`, but it doesn't read as nice.

`member_of: [Group1, Group2]` ?

Reply:
That would be the way to do it, yes, but could it be confusing for users?

Reply:
Another alternative: `changed Group.member_of` - basically renaming it from `Group.items`